18-filter-in-trees.md

```diff
@@ -0,0 +1,86 @@
+To do that, we have to extract the $\pi^-$ from `([D0 -> pi- K+]CC)` and combine it with all pions in `Phys/StdAllNoPIDsPions/Particles`.
+
+Using `FilterInTrees` is done in the same way we would use `FilterDesktop`:
+
+```python
+from Configurables import FilterInTrees
+from PhysSelPython.Wrappers import Selection, DataOnDemand
+
+decay_tree_location = '/Event/AllStreams/Phys/D2hhCompleteEventPromptDst2D2RSLine/Particles'
+d0_from_dst = FilterInTrees('d0_from_dst_filter', Code="DECTREE('[Charm -> pi- K+]CC')")
+d0_from_dst_sel = Selection("d0_from_dst_sel",
+                            Algorithm=d0_from_dst,
+                            RequiredSelections=[AutomaticData(Location=decay_tree_location)])
+pions_from_d0 = FilterInTrees('pions_from_d0_filter',Code="('pi+' == ABSID)")
+pions_from_d0_sel = Selection("pions_from_d0_sel",
+                              Algorithm=pions_from_d0,
+                              RequiredSelections=[d0_from_dst_sel])
+```
+
+The output of `pion_sel` is a container with all the pions coming from the $D^0$.
+
+> ## Question {.callout}
+> Do you see why we couldn't use something simple like
+> ```python
+pions_from_d0 = FilterInTrees('pions_from_d0_filter', Code="'pi+' == ABSID")
+```
+> ?
+
+Note how we had to do the process in two steps in order to avoid getting the soft pion from the $D^*$.
+Sometimes this makes things quite difficult, but almost all problems can be solved with a smart use of the `DECTREE` container in an intermediate step.
+
+> ## Selecting the soft photon {.challenge}
+> Can you find of a way of selecting the soft pion?
+> Hint: use the `FilterDecays` algorithm, in which you specify a decay descriptor as `Code`, marking the desired particle(s).
+
+The final step is easy, very similar to [building your own decay](https://lhcb.github.io/second-analysis-steps/01-building-decays.html):
+
+```python
+from Configurables import CombineParticles
+from PhysSelPython.Wrappers import Selection, DataOnDemand
+
+Pions = DataOnDemand('Phys/StdAllNoPIDsPions/Particles')
+rho = CombineParticles('rho_particles',
+                       DecayDescriptor=['rho(770)0 -> pi+ pi-'],
+                       CombinationCut="ADAMASS('rho(770)0') < 300*MeV",
+                       MotherCut='(VFASPF(VCHI2/VDOF)< 9)')
+rho_sel = Selection('rho_sel',
+                    Algorithm=rho,
+                    RequiredSelections=[pions_from_d0_sel, Pions])
+```
+
+Unfortunately, the `CombineParticles` example we just wrote is wrong, since it will actually build $\rho$ from all pions it gets as input, not using one from our `pions_from_d0` selection and one from `'Phys/StdAllNoPIDsPions/Particles'`.
+How to solve this?
+We have to get creative and use the tools at hand:
+for example, we could use `SubstitutePID` from the previous lesson to change the PID of the pions in the `pions_from_d0` selection to kaon and build `[rho(770)0 -> K+ pi-]CC` and then change again the PID of the kaon to a pion.
+Any other ideas?
```

Comment:
I took a more "advanced" approach here, showing the hardest case so they don't make the mistake. Maybe too much?